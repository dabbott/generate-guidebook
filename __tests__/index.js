const { Volume, createFsFromVolume } = require('memfs')
const scan = require('..')

function createFs(fileTree) {
  const volume = Volume.fromNestedJSON(fileTree, '/')
  const memfs = createFsFromVolume(volume)
  return memfs
}

describe('index', () => {
  it('converts an index file', () => {
    const fs = createFs({
      pages: {
        'index.mdx': '',
      },
    })

    const result = scan('/pages', undefined, fs)

    expect(result).toEqual({
      id: 0,
      file: 'index.mdx',
      slug: '',
      title: 'Index',
      subtitle: undefined,
      children: [],
      next: undefined,
      headings: [],
    })
  })

  it('reads frontmatter', () => {
    const fs = createFs({
      pages: {
        'index.mdx': `---
title: foo
subtitle: bar
---

# Content`,
      },
    })

    const result = scan('/pages', undefined, fs)

    expect(result).toEqual({
      id: 0,
      file: 'index.mdx',
      slug: '',
      title: 'foo',
      subtitle: 'bar',
      children: [],
      next: undefined,
      headings: [
        {
          level: 1,
          title: 'Content',
          url: '#content',
        },
      ],
    })
  })

  it('supports variables in frontmatter', () => {
    const fs = createFs({
      pages: {
        'index.mdx': `---
title: \${VARIABLE}
---

# Content`,
      },
    })

    const result = scan('/pages', { VARIABLE: 'Hello' }, fs)

    expect(result).toEqual({
      id: 0,
      file: 'index.mdx',
      slug: '',
      title: 'Hello',
      subtitle: undefined,
      children: [],
      next: undefined,
      headings: [
        {
          level: 1,
          title: 'Content',
          url: '#content',
        },
      ],
    })
  })

  it('handles order in config.json', () => {
    const fs = createFs({
      pages: {
        'index.mdx': '',
        'a.mdx': '',
        'b.mdx': '',
        'config.json': JSON.stringify({
          order: ['b', 'a'],
        }),
      },
    })

    const result = scan('/pages', undefined, fs)

    expect(result).toMatchSnapshot()
  })

  it('reads top-level directory', () => {
    const fs = createFs({
      pages: {
        'index.mdx': '',
        'a.mdx': '',
        'b.mdx': '',
      },
    })

    const result = scan('/pages', undefined, fs)

    expect(result).toMatchSnapshot()
  })

  it('reads nested directories', () => {
    const fs = createFs({
      pages: {
        'index.mdx': '',
        'a.mdx': '',
        a: {
          '1.mdx': '',
          '2.mdx': '',
        },
        'b.mdx': '',
        b: {
          '1.mdx': '',
        },
      },
    })

    const result = scan('/pages', undefined, fs)

    expect(result).toMatchSnapshot()
  })

  it('reads doubly nested directories', () => {
    const fs = createFs({
      pages: {
        'index.mdx': '',
        'a.mdx': '',
        a: {
          '1.mdx': '',
          '2.mdx': '',
          2: {
            'i.mdx': '',
            'j.mdx': '',
          },
        },
        'b.mdx': '',
        b: {
          '1.mdx': '',
        },
      },
    })

    const result = scan('/pages', undefined, fs)

    expect(result).toMatchSnapshot()
  })

  it('handles order in nested config.json', () => {
    const fs = createFs({
      pages: {
        'index.mdx': '',
        'a.mdx': '',
        a: {
          '1.mdx': '',
        },
        'hooks.mdx': '',
        hooks: {
          'custom_hooks.mdx': '',
          'usecontext.mdx': '',
          'useeffect.mdx': '',
          'usereducer.mdx': '',
          'useref.mdx': '',
          'usestate.mdx': `---
title: useState
---

# useState`,
          'config.json': JSON.stringify({
            order: [
              'usestate',
              'usereducer',
              'useeffect',
              'useref',
              'usecontext',
              'custom_hooks',
            ],
          }),
        },
        'config.json': JSON.stringify({
          order: ['a', 'hooks'],
        }),
      },
    })

    const result = scan('/pages', undefined, fs)

    expect(result).toMatchSnapshot()
  })

  describe('hidden', () => {
    it('handles hidden flag in frontmatter', () => {
      const fs = createFs({
        pages: {
          'index.mdx': '',
          'a.mdx': '',
          'b.mdx': `---
hidden: true
---`,
        },
      })

      const result = scan('/pages', undefined, fs)

      expect(result).toMatchSnapshot()
    })

    it('handles hidden directory in frontmatter', () => {
      const fs = createFs({
        pages: {
          'index.mdx': '',
          'a.mdx': `---
hidden: true
---`,
          a: {
            '1.mdx': '',
          },
        },
      })

      const result = scan('/pages', undefined, fs)

      expect(result).toMatchSnapshot()
    })

    it('handles hidden flag in frontmatter', () => {
      const fs = createFs({
        pages: {
          'index.mdx': '',
          'a.mdx': `---
hidden: TEST_VARIABLE
---`,
        },
      })

      const result = scan('/pages', { TEST_VARIABLE: true }, fs)

      expect(result).toMatchSnapshot()
    })

    it('handles hidden flag in frontmatter', () => {
      const fs = createFs({
        pages: {
          'index.mdx': '',
          'a.mdx': `---
hidden: UNDEFINED
---`,
        },
      })

      const result = scan('/pages', undefined, fs)

      expect(result).toMatchSnapshot()
    })
  })
})
