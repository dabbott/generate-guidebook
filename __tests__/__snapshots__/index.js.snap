// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index handles order in config.json 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "file": "b.mdx",
      "next": "a",
      "parent": "",
      "previous": "",
      "slug": "b",
      "subtitle": undefined,
      "title": "B",
    },
    Object {
      "children": Array [],
      "file": "a.mdx",
      "next": undefined,
      "parent": "",
      "previous": "b",
      "slug": "a",
      "subtitle": undefined,
      "title": "A",
    },
  ],
  "file": "index.mdx",
  "next": "b",
  "slug": "",
  "subtitle": undefined,
  "title": "Index",
}
`;

exports[`index reads doubly nested directories 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "file": "1.mdx",
          "next": "a/2",
          "parent": "a",
          "previous": "a",
          "slug": "a/1",
          "subtitle": undefined,
          "title": "1",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "file": "i.mdx",
              "next": "a/2/j",
              "parent": "a/2",
              "previous": "a/2",
              "slug": "a/2/i",
              "subtitle": undefined,
              "title": "I",
            },
            Object {
              "children": Array [],
              "file": "j.mdx",
              "next": "b",
              "parent": "a/2",
              "previous": "a/2/i",
              "slug": "a/2/j",
              "subtitle": undefined,
              "title": "J",
            },
          ],
          "file": "2.mdx",
          "next": "a/2/i",
          "parent": "a",
          "previous": "a/1",
          "slug": "a/2",
          "subtitle": undefined,
          "title": "2",
        },
      ],
      "file": "a.mdx",
      "next": "a/1",
      "parent": "",
      "previous": "",
      "slug": "a",
      "subtitle": undefined,
      "title": "A",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "file": "1.mdx",
          "next": undefined,
          "parent": "b",
          "previous": "b",
          "slug": "b/1",
          "subtitle": undefined,
          "title": "1",
        },
      ],
      "file": "b.mdx",
      "next": "b/1",
      "parent": "",
      "previous": "a",
      "slug": "b",
      "subtitle": undefined,
      "title": "B",
    },
  ],
  "file": "index.mdx",
  "next": "a",
  "slug": "",
  "subtitle": undefined,
  "title": "Index",
}
`;

exports[`index reads nested directories 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "file": "1.mdx",
          "next": "a/2",
          "parent": "a",
          "previous": "a",
          "slug": "a/1",
          "subtitle": undefined,
          "title": "1",
        },
        Object {
          "children": Array [],
          "file": "2.mdx",
          "next": "b",
          "parent": "a",
          "previous": "a/1",
          "slug": "a/2",
          "subtitle": undefined,
          "title": "2",
        },
      ],
      "file": "a.mdx",
      "next": "a/1",
      "parent": "",
      "previous": "",
      "slug": "a",
      "subtitle": undefined,
      "title": "A",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "file": "1.mdx",
          "next": undefined,
          "parent": "b",
          "previous": "b",
          "slug": "b/1",
          "subtitle": undefined,
          "title": "1",
        },
      ],
      "file": "b.mdx",
      "next": "b/1",
      "parent": "",
      "previous": "a",
      "slug": "b",
      "subtitle": undefined,
      "title": "B",
    },
  ],
  "file": "index.mdx",
  "next": "a",
  "slug": "",
  "subtitle": undefined,
  "title": "Index",
}
`;

exports[`index reads top-level directory 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "file": "a.mdx",
      "next": "b",
      "parent": "",
      "previous": "",
      "slug": "a",
      "subtitle": undefined,
      "title": "A",
    },
    Object {
      "children": Array [],
      "file": "b.mdx",
      "next": undefined,
      "parent": "",
      "previous": "a",
      "slug": "b",
      "subtitle": undefined,
      "title": "B",
    },
  ],
  "file": "index.mdx",
  "next": "a",
  "slug": "",
  "subtitle": undefined,
  "title": "Index",
}
`;
